### ISSUE TITLE
AdaBoost

### ISSUE THREAD

**Issue by YuusukeTabata** [YuusukeTabata/friendly-bassoon#1]
ーアンサンブル法
　ーブースティング
　　ーAdaBoost
- [概要] 
　ランダム予測よりもわずかに優れている程度の、弱学習器とも呼ばれるベース分類器を複数用意したら、重み付けられたデータセットを「逐次的に」学習させる。直前のベース分類器の性能に依存する重み係数で重み付けが更新されたデータセットによってすべてのベース分類器の学習が完了したら、各々の予測結果を重み付き多数決投票スキームにより結合し、最終的な予測とする。
- [アルゴリズム （ここではSAMME）] 
二値のクラスラベル分類を想定すると、以下のステップに要約される。

1.  N個のデータに1/Nずつ重み付けを行う（＝均等に重み付ける）。
2. 分類器y(x)がm個ある＝全体でm回実行するブースティングにおいて、ベース分類器y1(x)がまず学習する。

>【source code】
641`"""Compute the decision function of ``X``.
        Parameters
        ----------
        X : {array-like, sparse matrix} of shape = [n_samples, n_features]
            The training input samples. Sparse matrix can be CSC, CSR, COO,
            DOK, or LIL. DOK and LIL are converted to CSR.
        Returns
        -------
        score : array, shape = [n_samples, k]
            The decision function of the input samples. The order of
            outputs is the same of that of the `classes_` attribute.
            Binary classification is a special cases with ``k == 1``,
            otherwise ``k==n_classes``. For binary classification,
            values closer to -1 or 1 mean more like the first or second
            class in ``classes_``, respectively.
        """`
> 

       671 else: # self.algorithm == "SAMME"
                   pred = sum((estimator.predict(X) == classes).T * w
                            for estimator, w in zip(self.estimators_,
                                                      self.estimator_weights_))
              pred /= self.estimator_weights_.sum()
              if n_classes == 2:
                      pred[:, 0] *= -1
                      return pred.sum(axis=1)
              return pred
3.  y1(x)の予測結果y^1と真のクラスラベルを照合して、均等重み付けされたなかでの誤分類率εを計算する。
　　ε1=**w** 1(y^1≠y1)
4.  重みの更新に用いる係数α1を計算する。
　　α1=0.5 log 1-ε1/ε1
>【source code】

    572   estimator_weight = self.learning_rate * (np.log((1. - estimator_error) / estimator_error) 
              +np.log(n_classes - 1.))

5.  重みを更新する。
　**w** =**w** 1 exp(-α1・y^1・y1)
　　y^1==y1ならば、y^1・y1の符号は正になる。→αはマイナス符号＝重みを減らす。
　　y^1！=y1ならば、y^1・y1の符号は負になる。→αはマイナス符号＝重みを増やす。
>【source code】
 'Only boost positive weights'

    579 sample_weight *= np.exp(estimator_weight * incorrect *
                              ((sample_weight > 0) |(estimator_weight < 0)))

6.  重みが更新されたら、正規化して合計１になるようにする。
>【source code】
736` """Predict class probabilities for X.
        The predicted class probabilities of an input sample is computed as
        the weighted mean predicted class probabilities of the classifiers
        in the ensemble.`

    766 def predict_proba(self, X):
             else:   # self.algorithm == "SAMME"
            proba = sum(estimator.predict_proba(X) * w
                        for estimator, w in zip(self.estimators_,
                                                self.estimator_weights_))
        proba /= self.estimator_weights_.sum()
        proba = np.exp((1. / (n_classes - 1)) * proba)
        normalizer = proba.sum(axis=1)[:, np.newaxis]
        normalizer[normalizer == 0.0] = 1.0
        proba /= normalizer

        return proba

以上でy1(x)の学習が完了し、更新された重み付きデータセットをy2(x)へ引き継ぎ、y2(x)の学習が開始される。以降、y3(x),...ym(x)まで 2. から 6. を繰り返す。

7.  ym(x)の学習が完了したら、最終的な予測＝全体の出力として、各々の予測結果を重み付き多数決投票スキームにより結合する。多数決の際には、重み係数αを再活用して、精度の良かった分類器に対して重みを大きくするインセンティブを持たせる。
>【source code】
585`"""Predict classes for X.
        The predicted class of an input sample is computed as the weighted mean
        prediction of the classifiers in the ensemble.
        Parameters
        ----------
        X : {array-like, sparse matrix} of shape = [n_samples, n_features]
            The training input samples. Sparse matrix can be CSC, CSR, COO,
            DOK, or LIL. DOK and LIL are converted to CSR.
        Returns
        -------
        y : array of shape = [n_samples]
            The predicted classes.
        """`
> 

      def predict(self, X):
      pred = self.decision_function(X)
      if self.n_classes_ == 2:
           return self.classes_.take(pred > 0, axis=0)
      return self.classes_.take(np.argmax(pred, axis=1), axis=0)

- [ ] ada

**Comment by YuusukeTabata** [comment]
ブースティングにおいて最も一般的な実装であるAdaBoost。
